In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import json
from collections import Counter

### Inspect the train, test and submission dataset

In [2]:
train= pd.read_json(r'C:\Users\HP\Desktop\codellc\CovidVaccine\train\train.json',  lines= True)

In [3]:
train.head()

index            id                                           sequence  \
0      0  id_001f94081  GGAAAAGCUCUAAUAACAGGAGACUAGGACUACGUAUUUCUAGGUA...   
1      1  id_0049f53ba  GGAAAAAGCGCGCGCGGUUAGCGCGCGCUUUUGCGCGCGCUGUACC...   
2      2  id_006f36f57  GGAAAGUGCUCAGAUAAGCUAAGCUCGAAUAGCAAUCGAAUAGAAU...   
3      3  id_0082d463b  GGAAAAGCGCGCGCGCGCGCGCGAAAAAGCGCGCGCGCGCGCGCGC...   
4      4  id_0087940f4  GGAAAAUAUAUAAUAUAUUAUAUAAAUAUAUUAUAGAAGUAUAAUA...   

                                           structure  \
0  .....((((((.......)))).)).((.....((..((((((......   
1  .....(((((((((((((((((((((((....)))))))))).)))...   
2  .....((((.((.....((((.(((.....)))..((((......)...   
3  ......((((((((((((((((......))))))))))))))))((...   
4  .....(((((((.((((((((((((.(((((((((....)))))))...   

                                 predicted_loop_type  signal_to_noise  \
0  EEEEESSSSSSHHHHHHHSSSSBSSXSSIIIIISSIISSSSSSHHH...            6.894   
1  EEEEESSSSSSSSSSSSSSSSSSSSSSSHHHHSSSSSSSSSSBSSS...            0.193   
2  EEEEESSSSISSIIIIISSSSMSSSHHHHHSSSMMSSSSHHHHHHS...            8.800   
3  EEEEEESSSSSSSSSSSSSSSSHHHHHHSSSSSSSSSSSSSSSSSS...            0.104   
4  EEEEESSSSSSSBSSSSSSSSSSSSBSSSSSSSSSHHHHSSSSSSS...            0.423   

   SN_filter  seq_length  seq_scored  \
0          1         107          68   
1          0         107          68   
2          1         107          68   
3          0         107          68   
4          0         107          68   

                                    reactivity_error  \
0  [0.1359, 0.20700000000000002, 0.1633, 0.1452, ...   
1  [2.8272, 2.8272, 2.8272, 4.7343, 2.5676, 2.567...   
2  [0.0931, 0.13290000000000002, 0.11280000000000...   
3  [3.5229, 6.0748, 3.0374, 3.0374, 3.0374, 3.037...   
4  [1.665, 2.1728, 2.0041, 1.2405, 0.620200000000...   

                                   deg_error_Mg_pH10  \
0  [0.26130000000000003, 0.38420000000000004, 0.1...   
1  [73705.3985, 73705.3985, 73705.3985, 73705.398...   
2  [0.1365, 0.2237, 0.1812, 0.1333, 0.1148, 0.160...   
3  [73705.3985, 73705.3985, 73705.3985, 73705.398...   
4  [4.2139, 3.9637000000000002, 3.2467, 2.4716, 1...   

                                      deg_error_pH10  \
0  [0.2631, 0.28600000000000003, 0.0964, 0.1574, ...   
1  [10.1986, 9.2418, 5.0933, 5.0933, 5.0933, 5.09...   
2  [0.17020000000000002, 0.178, 0.111, 0.091, 0.0...   
3  [11.8007, 12.7566, 5.7733, 5.7733, 5.7733, 5.7...   
4  [3.0942, 3.015, 2.1212, 2.0552, 0.881500000000...   

                                    deg_error_Mg_50C  \
0  [0.1501, 0.275, 0.0947, 0.18660000000000002, 0...   
1  [16.6174, 13.868, 8.1968, 8.1968, 8.1968, 8.19...   
2  [0.1033, 0.1464, 0.1126, 0.09620000000000001, ...   
3  [121286.7181, 121286.7182, 121286.7181, 121286...   
4  [2.6717, 2.4818, 1.9919, 2.5484999999999998, 1...   

                                       deg_error_50C  \
0  [0.2167, 0.34750000000000003, 0.188, 0.2124, 0...   
1  [15.4857, 7.9596, 13.3957, 5.8777, 5.8777, 5.8...   
2  [0.14980000000000002, 0.1761, 0.1517, 0.116700...   
3  [15.3995, 8.1124, 7.7824, 7.7824, 7.7824, 7.78...   
4  [1.3285, 3.6173, 1.3057, 1.3021, 1.1507, 1.150...   

                                          reactivity  \
0  [0.3297, 1.5693000000000001, 1.1227, 0.8686, 0...   
1  [0.0, 0.0, 0.0, 2.2965, 0.0, 0.0, 0.0, 0.0, 0....   
2  [0.44820000000000004, 1.4822, 1.1819, 0.743400...   
3  [0.0, 2.2399, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0....   
4  [0.8267, 2.6577, 2.8481, 0.40090000000000003, ...   

                                         deg_Mg_pH10  \
0  [0.7556, 2.983, 0.2526, 1.3789, 0.637600000000...   
1  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
2  [0.2504, 1.4021, 0.9804, 0.49670000000000003, ...   
3  [0.0, -0.5083, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0...   
4  [2.1058, 3.138, 2.5437000000000003, 1.0932, 0....   

                                            deg_pH10  \
0  [2.3375, 3.5060000000000002, 0.3008, 1.0108, 0...   
1  [4.947, 4.4523, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...

In [4]:
train.shape

(2400, 19)

In [5]:
test= pd.read_json(r'C:\Users\HP\Desktop\codellc\CovidVaccine\test\test.json',  lines= True)

In [6]:
test.head()

index            id                                           sequence  \
0      0  id_00073f8be  GGAAAAGUACGACUUGAGUACGGAAAACGUACCAACUCGAUUAAAA...   
1      1  id_000ae4237  GGAAACGGGUUCCGCGGAUUGCUGCUAAUAAGAGUAAUCUCUAAAU...   
2      2  id_00131c573  GGAAAACAAAACGGCCUGGAAGACGAAGGAAUUCGGCGCGAAGGCC...   
3      3  id_00181fd34  GGAAAGGAUCUCUAUCGAAGGAUAGAGAUCGCUCGCGACGGCACGA...   
4      4  id_0020473f7  GGAAACCCGCCCGCGCCCGCCCGCGCUGCUGCCGUGCCUCCUCUCC...   

                                           structure  \
0  ......((((((((((.(((((.....))))))))((((((((......   
1  .....((((..((((((...(((((.....((((....)))).......   
2  ...........((.(((.(.(..((..((..((((...))))..))...   
3  ......((((((((((....))))))))))((((((..((.(((.....   
4  .....(((((((((((((((((((((((((((((((((((((((((...   

                                 predicted_loop_type  seq_length  seq_scored  
0  EEEEEESSSSSSSSSSBSSSSSHHHHHSSSSSSSSSSSSSSSSHHH...         107          68  
1  EEEEESSSSIISSSSSSIIISSSSSIIIIISSSSHHHHSSSSIIII...         130          91  
2  EEEEEEEEEEESSISSSISISIISSIISSIISSSSHHHSSSSIISS...         107          68  
3  EEEEEESSSSSSSSSSHHHHSSSSSSSSSSSSSSSSIISSISSSHH...         107          68  
4  EEEEESSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS...         130          91

In [7]:
test.shape

(3634, 7)

In [8]:
submission = pd.read_csv(r'C:/Users/HP/Desktop/codellc/CovidVaccine/sample_submission/sample_submission.csv')

In [9]:
submission.head()

id_seqpos  reactivity  deg_Mg_pH10  deg_pH10  deg_Mg_50C  deg_50C
0  id_00073f8be_0         0.0          0.0       0.0         0.0      0.0
1  id_00073f8be_1         0.0          0.0       0.0         0.0      0.0
2  id_00073f8be_2         0.0          0.0       0.0         0.0      0.0
3  id_00073f8be_3         0.0          0.0       0.0         0.0      0.0
4  id_00073f8be_4         0.0          0.0       0.0         0.0      0.0

In [10]:
print(submission.shape)

(457953, 6)


In [11]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2400 entries, 0 to 2399
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   index                2400 non-null   int64  
 1   id                   2400 non-null   object 
 2   sequence             2400 non-null   object 
 3   structure            2400 non-null   object 
 4   predicted_loop_type  2400 non-null   object 
 5   signal_to_noise      2400 non-null   float64
 6   SN_filter            2400 non-null   int64  
 7   seq_length           2400 non-null   int64  
 8   seq_scored           2400 non-null   int64  
 9   reactivity_error     2400 non-null   object 
 10  deg_error_Mg_pH10    2400 non-null   object 
 11  deg_error_pH10       2400 non-null   object 
 12  deg_error_Mg_50C     2400 non-null   object 
 13  deg_error_50C        2400 non-null   object 
 14  reactivity           2400 non-null   object 
 15  deg_Mg_pH10          2400 non-null   o

In [12]:
test['seq_scored'][test['seq_scored'] == 68].count()

629

In [13]:
sample = train[train['id'] == 'id_001f94081']
sample

index            id                                           sequence  \
0      0  id_001f94081  GGAAAAGCUCUAAUAACAGGAGACUAGGACUACGUAUUUCUAGGUA...   

                                           structure  \
0  .....((((((.......)))).)).((.....((..((((((......   

                                 predicted_loop_type  signal_to_noise  \
0  EEEEESSSSSSHHHHHHHSSSSBSSXSSIIIIISSIISSSSSSHHH...            6.894   

   SN_filter  seq_length  seq_scored  \
0          1         107          68   

                                    reactivity_error  \
0  [0.1359, 0.20700000000000002, 0.1633, 0.1452, ...   

                                   deg_error_Mg_pH10  \
0  [0.26130000000000003, 0.38420000000000004, 0.1...   

                                      deg_error_pH10  \
0  [0.2631, 0.28600000000000003, 0.0964, 0.1574, ...   

                                    deg_error_Mg_50C  \
0  [0.1501, 0.275, 0.0947, 0.18660000000000002, 0...   

                                       deg_error_50C  \
0  [0.2167, 0.34750000000000003, 0.188, 0.2124, 0...   

                                          reactivity  \
0  [0.3297, 1.5693000000000001, 1.1227, 0.8686, 0...   

                                         deg_Mg_pH10  \
0  [0.7556, 2.983, 0.2526, 1.3789, 0.637600000000...   

                                            deg_pH10  \
0  [2.3375, 3.5060000000000002, 0.3008, 1.0108, 0...   

                                          deg_Mg_50C  \
0  [0.35810000000000003, 2.9683, 0.2589, 1.4552, ...   

                                             deg_50C  
0  [0.6382, 3.4773, 0.9988, 1.3228, 0.78770000000...

1. sequence: This refers to an RNA sequence with 4 bases namely Adenine, Guanine, Cytosine and Uracil represented with A, G, C, U respectively

2. Structure: This is an array of special characters that shows the nature of base-pairing rule that exists between the bases of the RNA

3. Predicted loop type: This describes the structural context of each character in sequence. Loop types assigned by bpRNA from Vienna RNAfold 2 structure. From the bpRNA_documentation: S: paired "Stem", M: Multiloop, I: Internal loop, B: Bulge, H: Hairpin loop, E: dangling End, X: eXternal loop.

In [14]:
sample['sequence'].values[0]

'GGAAAAGCUCUAAUAACAGGAGACUAGGACUACGUAUUUCUAGGUAACUGGAAUAACCCAUACCAGCAGUUAGAGUUCGCUCUAACAAAAGAAACAACAACAACAAC'

### Looking at the dataframe columns, it is evident that the sequence has a known length and a known score which are represented by 'seq_length' and 'seq_score' respectively. To know the value for each of the bases of the RNA sequence, we use the Counter module

In [15]:
Counter(sample['sequence'].values[0])

Counter({'G': 19, 'A': 45, 'C': 23, 'U': 20})

   ### Addition of all the values of the nucleotide bases(G,A,C,U) gives a total of 107 which is equal to the 'seq_length'

In [16]:
sample['structure'].values[0]

'.....((((((.......)))).)).((.....((..((((((....))))))..)).....))....(((((((....))))))).....................'

### In the structure column, there are 3 special characters and each of these characters have their assigned values and meanings. For instance, (....) means that base 0 is paired to base 5 and bases 1-4 are unpaired. The special characters include
1. .
2. (
3. )



In [17]:
Counter(sample['structure'].values[0])

Counter({'.': 61, '(': 23, ')': 23})

In [18]:
61 + 23 +23

107

### Predicted loop type: Describes the structural context of each character in sequence. Loop types assigned by bpRNA from Vienna RNAfold 2 structure. From the bpRNA_documentation: S: paired "Stem" M: Multiloop I: Internal loop B: Bulge H: Hairpin loop E: dangling End X: eXternal loop

In [19]:
sample['predicted_loop_type'].values[0]

'EEEEESSSSSSHHHHHHHSSSSBSSXSSIIIIISSIISSSSSSHHHHSSSSSSIISSIIIIISSXXXXSSSSSSSHHHHSSSSSSSEEEEEEEEEEEEEEEEEEEEE'

In [20]:
Counter(sample['predicted_loop_type'].values[0])

Counter({'E': 26, 'S': 46, 'H': 15, 'B': 1, 'X': 5, 'I': 14})

In [21]:
train['seq_length'].value_counts()

107    2400
Name: seq_length, dtype: int64

In [22]:
test['seq_length'].value_counts()

130    3005
107     629
Name: seq_length, dtype: int64

## Filtering the data 

From the "Data" tab of this competition, there are important things to note:

1. At the beginning of the competition, Stanford scientists have data on 3029 RNA sequences of length 107. For technical reasons, measurements cannot be carried out on the final bases of these RNA sequences, so we have experimental data (ground truth) in 5 conditions for the first 68 bases.

2. We have split out 629 of these 3029 sequences for a public test set to allow for continuous evaluation through the competition. This leaves of 2400 for the training set.

3. In the 3634 data set present in "test.json", 629 are the 107-base sequence, while the remaining 3005 is the Private new MRNAs.

4. Measurements can be done only on the first 68 points of sequences.

5. Data was filtered and measurements were done based on 5 conditions (reactivity, deg_Mg_pH10, deg_pH10, deg_Mg_50C, deg_50C).

Filtering was done using the following criteria:

1. Minimum value across all 5 conditions must be greater than -0.5.

2. Mean signal/noise across all 5 conditions must be greater than 1.0. "Signal/noise is defined as mean( measurement value over 68 nts )/mean( statistical error in measurement value over 68 nts)"

3. To help ensure sequence diversity, the resulting sequences were clustered into clusters with less than 50% sequence similarity, and the 629 test set sequences were chosen from clusters with 3 or fewer members. That is, any sequence in the test set should be sequence similar to at most 2 other sequences.


The training data was not filtered, therefore we need to begin filtering from the training dat set using the above criteria. 

### What are our targets? For the purpose of scoring, only the following needs to be gotten `reactivity`, `deg_Mg_pH10` and `deg_Mg_50C`.

In [23]:
train.head(2)

index            id                                           sequence  \
0      0  id_001f94081  GGAAAAGCUCUAAUAACAGGAGACUAGGACUACGUAUUUCUAGGUA...   
1      1  id_0049f53ba  GGAAAAAGCGCGCGCGGUUAGCGCGCGCUUUUGCGCGCGCUGUACC...   

                                           structure  \
0  .....((((((.......)))).)).((.....((..((((((......   
1  .....(((((((((((((((((((((((....)))))))))).)))...   

                                 predicted_loop_type  signal_to_noise  \
0  EEEEESSSSSSHHHHHHHSSSSBSSXSSIIIIISSIISSSSSSHHH...            6.894   
1  EEEEESSSSSSSSSSSSSSSSSSSSSSSHHHHSSSSSSSSSSBSSS...            0.193   

   SN_filter  seq_length  seq_scored  \
0          1         107          68   
1          0         107          68   

                                    reactivity_error  \
0  [0.1359, 0.20700000000000002, 0.1633, 0.1452, ...   
1  [2.8272, 2.8272, 2.8272, 4.7343, 2.5676, 2.567...   

                                   deg_error_Mg_pH10  \
0  [0.26130000000000003, 0.38420000000000004, 0.1...   
1  [73705.3985, 73705.3985, 73705.3985, 73705.398...   

                                      deg_error_pH10  \
0  [0.2631, 0.28600000000000003, 0.0964, 0.1574, ...   
1  [10.1986, 9.2418, 5.0933, 5.0933, 5.0933, 5.09...   

                                    deg_error_Mg_50C  \
0  [0.1501, 0.275, 0.0947, 0.18660000000000002, 0...   
1  [16.6174, 13.868, 8.1968, 8.1968, 8.1968, 8.19...   

                                       deg_error_50C  \
0  [0.2167, 0.34750000000000003, 0.188, 0.2124, 0...   
1  [15.4857, 7.9596, 13.3957, 5.8777, 5.8777, 5.8...   

                                          reactivity  \
0  [0.3297, 1.5693000000000001, 1.1227, 0.8686, 0...   
1  [0.0, 0.0, 0.0, 2.2965, 0.0, 0.0, 0.0, 0.0, 0....   

                                         deg_Mg_pH10  \
0  [0.7556, 2.983, 0.2526, 1.3789, 0.637600000000...   
1  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   

                                            deg_pH10  \
0  [2.3375, 3.5060000000000002, 0.3008, 1.0108, 0...   
1  [4.947, 4.4523, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   

                                          deg_Mg_50C  \
0  [0.35810000000000003, 2.9683, 0.2589, 1.4552, ...   
1  [4.8511, 4.0426, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...   

                                             deg_50C  
0  [0.6382, 3.4773, 0.9988, 1.3228, 0.78770000000...  
1  [7.6692, 0.0, 10.9561, 0.0, 0.0, 0.0, 0.0, 0.0...

### Inspecting the training set, public set and private set

In [24]:
train['seq_scored'].value_counts()

68    2400
Name: seq_scored, dtype: int64

In [25]:
test['seq_scored'].value_counts()

91    3005
68     629
Name: seq_scored, dtype: int64

From the above two lines of code, it is evident that both the train and test data have different values. In the test dataset, row with 68 is the public dataset while the one with 91 is the private dataset.

### Feature Engineering

In [26]:
train_data = []
for mRNA_id in train['id'].unique():
    sample_data = train.loc[train['id'] == mRNA_id]
    for i in range(68):
        sample_tuple = (sample_data['id'].values[0], sample_data['sequence'].values[0][:3],
                        sample_data['structure'].values[0][i], sample_data['predicted_loop_type'].values[0][i],
                        sample_data['reactivity'].values[0][i], sample_data['reactivity_error'].values[0][i],
                        sample_data['deg_Mg_pH10'].values[0][i], sample_data['deg_error_Mg_pH10'].values[0][i],
                        sample_data['deg_pH10'].values[0][i], sample_data['deg_error_pH10'].values[0][i],
                        sample_data['deg_Mg_50C'].values[0][i], sample_data['deg_error_Mg_50C'].values[0][i],
                        sample_data['deg_50C'].values[0][i], sample_data['deg_error_50C'].values[0][i])
        train_data.append(sample_tuple)

In [27]:
train_data = pd.DataFrame(train_data, columns=['id', 'sequence', 'structure', 'predicted_loop_type', 'reactivity', 'reactivity_error', 'deg_Mg_pH10', 'deg_error_Mg_pH10',
                                  'deg_pH10', 'deg_error_pH10', 'deg_Mg_50C', 'deg_error_Mg_50C', 'deg_50C', 'deg_error_50C'])
train_data.head()

id sequence structure predicted_loop_type  reactivity  \
0  id_001f94081      GGA         .                   E      0.3297   
1  id_001f94081      GGA         .                   E      1.5693   
2  id_001f94081      GGA         .                   E      1.1227   
3  id_001f94081      GGA         .                   E      0.8686   
4  id_001f94081      GGA         .                   E      0.7217   

   reactivity_error  deg_Mg_pH10  deg_error_Mg_pH10  deg_pH10  deg_error_pH10  \
0            0.1359       0.7556             0.2613    2.3375          0.2631   
1            0.2070       2.9830             0.3842    3.5060          0.2860   
2            0.1633       0.2526             0.1372    0.3008          0.0964   
3            0.1452       1.3789             0.2531    1.0108          0.1574   
4            0.1314       0.6376             0.1798    0.2635          0.1000   

   deg_Mg_50C  deg_error_Mg_50C  deg_50C  deg_error_50C  
0      0.3581            0.1501   0.6382         0.2167  
1      2.9683            0.2750   3.4773         0.3475  
2      0.2589            0.0947   0.9988         0.1880  
3      1.4552            0.1866   1.3228         0.2124  
4      0.7244            0.1369   0.7877         0.1703

In [28]:
train_data.sort_values('reactivity_error')

id sequence structure predicted_loop_type  reactivity  \
54443   id_5429cb841      GGA         )                   S      0.0000   
26991   id_2a3308237      GGA         )                   S      0.0016   
54421   id_5429cb841      GGA         (                   S     -0.0009   
26993   id_2a3308237      GGA         )                   S      0.0024   
85089   id_8622860d7      GGA         (                   S      0.0000   
...              ...      ...       ...                 ...         ...   
101750  id_a1719ebbc      GGA         (                   S      0.0000   
101751  id_a1719ebbc      GGA         (                   S      0.0000   
101752  id_a1719ebbc      GGA         (                   S      0.0000   
101754  id_a1719ebbc      GGA         (                   S      0.0000   
24034   id_251bb7979      GGA         .                   I      0.0000   

        reactivity_error  deg_Mg_pH10  deg_error_Mg_pH10  deg_pH10  \
54443             0.0019       0.0331             0.0173    0.0608   
26991             0.0022       0.0426             0.0066    0.0336   
54421             0.0029       0.0250             0.0234    0.0056   
26993             0.0031       0.0950             0.0094    0.0641   
85089             0.0031       0.0102             0.0204    0.0051   
...                  ...          ...                ...       ...   
101750       146151.2250       0.0000        104235.1742    0.0000   
101751       146151.2250       0.0000        104235.1742    0.0000   
101752       146151.2250       0.0000        104235.1742    0.0000   
101754       146151.2250       0.0000        104235.1742    0.0000   
24034        146151.2250       0.0000        104235.1742    0.0000   

        deg_error_pH10  deg_Mg_50C  deg_error_Mg_50C  deg_50C  deg_error_50C  
54443           0.0178      0.0072            0.0066   0.0233         0.0135  
26991           0.0097      0.0481            0.0089   0.0377         0.0115  
54421           0.0101      0.0019            0.0065   0.0071         0.0113  
26993           0.0131      0.0761            0.0110   0.0547         0.0138  
85089           0.0105      0.0115            0.0106   0.0057         0.0117  
...                ...         ...               ...      ...            ...  
101750     314833.5695      0.0000       171525.3217   0.0000    271158.6040  
101751     314833.5695      0.0000       171525.3217   0.0000    271158.6040  
101752     314833.5695      0.0000       171525.3217   0.0000    271158.6040  
101754     314833.5695      0.0000       171525.3217   0.0000    271158.6040  
24034      314833.5695      0.0000       171525.3217   0.0000    271158.6040  

[163200 rows x 14 columns]

### Inspect the data again

In [29]:
train.loc[train['id'] == 'id_251bb7979']

index            id                                           sequence  \
353    353  id_251bb7979  GGAAAGGCGGGCGGGCGGGCGGGCGGGCGGGCGGGAAAACGGGCGG...   

                                             structure  \
353  .....(.(.(.(.(.(.(.(.(.(.(.(.(.(.(.....).).).)...   

                                   predicted_loop_type  signal_to_noise  \
353  EEEEESISISISISISISISISISISISISISISHHHHHSISISIS...              0.0   

     SN_filter  seq_length  seq_scored  \
353          0         107          68   

                                      reactivity_error  \
353  [146151.225, 146151.225, 146151.225, 146151.22...   

                                     deg_error_Mg_pH10  \
353  [104235.1742, 104235.1742, 104235.1742, 104235...   

                                        deg_error_pH10  \
353  [314833.5695, 314833.5695, 314833.5695, 314833...   

                                      deg_error_Mg_50C  \
353  [171525.3217, 171525.3217, 171525.3217, 171525...   

                                         deg_error_50C  \
353  [271158.604, 271158.604, 271158.604, 271158.60...   

                                            reactivity  \
353  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   

                                           deg_Mg_pH10  \
353  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   

                                              deg_pH10  \
353  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   

                                            deg_Mg_50C  \
353  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   

                                               deg_50C  
353  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...

From the above inspection, it is evident that there are some columns which contains 0 as values. It means that they do not fall within the set criteria or they are not among the first 68 set of sequences which will require further work

In [30]:
test_data = []
for mRNA_id in test['id'].unique():
    sample_data = test.loc[test['id'] == mRNA_id]
    for i in range(sample_data['seq_scored'].values[0]):
        sample_tuple = (sample_data['id'].values[0] + f'_{i}', sample_data['sequence'].values[0][:3],
                        sample_data['structure'].values[0][i], sample_data['predicted_loop_type'].values[0][i])
        test_data.append(sample_tuple)

In [31]:
test_data = pd.DataFrame(test_data, columns=['id', 'sequence', 'structure', 'predicted_loop_type'])
test_data.head()

id sequence structure predicted_loop_type
0  id_00073f8be_0      GGA         .                   E
1  id_00073f8be_1      GGA         .                   E
2  id_00073f8be_2      GGA         .                   E
3  id_00073f8be_3      GGA         .                   E
4  id_00073f8be_4      GGA         .                   E

In [32]:
train_data.groupby(['sequence', 'structure', 'predicted_loop_type'])['reactivity'].mean().reset_index().head()

sequence structure predicted_loop_type  reactivity
0      GGA         (                   S    0.181767
1      GGA         )                   S    0.176390
2      GGA         .                   B    0.628836
3      GGA         .                   E    0.562947
4      GGA         .                   H    0.677434

### Merge the data onto the test data set

In [33]:
test_data = pd.merge(test_data, train_data.groupby(['sequence', 'structure', 'predicted_loop_type'])['reactivity'].mean().reset_index(),
                     on=['sequence', 'structure', 'predicted_loop_type'])
test_data = pd.merge(test_data, train_data.groupby(['sequence', 'structure', 'predicted_loop_type'])['deg_Mg_pH10'].mean().reset_index(),
                     on=['sequence', 'structure', 'predicted_loop_type'])
test_data = pd.merge(test_data, train_data.groupby(['sequence', 'structure', 'predicted_loop_type'])['deg_pH10'].mean().reset_index(),
                     on=['sequence', 'structure', 'predicted_loop_type'])
test_data = pd.merge(test_data, train_data.groupby(['sequence', 'structure', 'predicted_loop_type'])['deg_Mg_50C'].mean().reset_index(),
                     on=['sequence', 'structure', 'predicted_loop_type'])
test_data = pd.merge(test_data, train_data.groupby(['sequence', 'structure', 'predicted_loop_type'])['deg_50C'].mean().reset_index(),
                     on=['sequence', 'structure', 'predicted_loop_type'])

In [34]:
test_data.head()

id sequence structure predicted_loop_type  reactivity  \
0  id_00073f8be_0      GGA         .                   E    0.562947   
1  id_00073f8be_1      GGA         .                   E    0.562947   
2  id_00073f8be_2      GGA         .                   E    0.562947   
3  id_00073f8be_3      GGA         .                   E    0.562947   
4  id_00073f8be_4      GGA         .                   E    0.562947   

   deg_Mg_pH10  deg_pH10  deg_Mg_50C   deg_50C  
0     0.586123  0.801656    0.641628  0.658497  
1     0.586123  0.801656    0.641628  0.658497  
2     0.586123  0.801656    0.641628  0.658497  
3     0.586123  0.801656    0.641628  0.658497  
4     0.586123  0.801656    0.641628  0.658497

### View the submission format

In [35]:
submission.head()

id_seqpos  reactivity  deg_Mg_pH10  deg_pH10  deg_Mg_50C  deg_50C
0  id_00073f8be_0         0.0          0.0       0.0         0.0      0.0
1  id_00073f8be_1         0.0          0.0       0.0         0.0      0.0
2  id_00073f8be_2         0.0          0.0       0.0         0.0      0.0
3  id_00073f8be_3         0.0          0.0       0.0         0.0      0.0
4  id_00073f8be_4         0.0          0.0       0.0         0.0      0.0

In [36]:
submission.shape, test_data.shape

((457953, 6), (316227, 9))

Rename the `id` column and do a left merging so that other columns not present in the submission column will be filled will null and then can be dropped.

In [37]:
submission_1 = pd.merge(submission[['id_seqpos']], test_data, left_on='id_seqpos', right_on='id', how='left').drop(['id', 'sequence', 'structure', 'predicted_loop_type'], axis=1)
submission_1.head()

id_seqpos  reactivity  deg_Mg_pH10  deg_pH10  deg_Mg_50C   deg_50C
0  id_00073f8be_0    0.562947     0.586123  0.801656    0.641628  0.658497
1  id_00073f8be_1    0.562947     0.586123  0.801656    0.641628  0.658497
2  id_00073f8be_2    0.562947     0.586123  0.801656    0.641628  0.658497
3  id_00073f8be_3    0.562947     0.586123  0.801656    0.641628  0.658497
4  id_00073f8be_4    0.562947     0.586123  0.801656    0.641628  0.658497

In [38]:
submission_1['reactivity'].isnull().sum()

141726

In [39]:
submission_1[['reactivity', 'deg_Mg_pH10', 'deg_pH10', 'deg_Mg_50C', 'deg_50C']] = submission_1[['reactivity', 'deg_Mg_pH10', 'deg_pH10', 'deg_Mg_50C', 'deg_50C']].fillna(0) * 0.9

In [40]:
submission_1.to_csv('submission.csv', index=False)